### DISCUSSION TITLE
Customising http response for listener

### DISCUSSION THREAD

**Discussion by scrumgelist** [nextgenhealthcare/connect#5587] [Q&A] [answered]
I'm close to what I want, but really struggling with the last bit...

I'm sure it is simple and when one of you good people tell me how to do it, my head and palm will collide.  None-the-less I have tried and reasearched and, well I'm guess I am missing something too obvious for anyone to have thought to explain.

I need to customise the response returned to an http listener.

I am able to pass back the response from the destination (a soap accepting web service) and set the http status, but have not yet worked out how to set the status message.  (Testing in postman, I get the correct status and see the soap response in the response content, but the status message is just a repeat of the response status.)

![image](https://user-images.githubusercontent.com/99478435/207293289-6da4a868-2494-4e72-a33f-802e4a08bc22.png)

In the above example I would really like to see "540 Patient not found" rather than "540 540"

As the incoming message moves through the channel I track events and set the required responce status code in a channelMap variable and pass this back by placing this (as a velocity variable) in the response status code field of the source http listener settings form.

![image](https://user-images.githubusercontent.com/99478435/207292120-6bfadc21-e3b9-443e-80da-4221d3962a46.png)

I also track the status message I would like to send with the status code.  I have tried combining these two in variously formatted ways, flat strings, JSON.stringify(customeStatusJSObject), even XML as that is what I get returned from the web service but all attempts have so far failed.

Please, what obvious thing am I not doing?

**Comment by jonbartels** [comment]
https://docs.nextgen.com/bundle/Mirth_User_Guide_40/page/connect/connect/topics/c_Response_MapResponse_Map_connect_ug.html

**Reply by jonbartels** [reply]
It may help to reduce this to an MCVE* which is just an HTTP listener spitting back the respose code and string you want. That'll separate this task from your bigger SOAP, patient lookup interactions.

*- minimal complete verifable example

**Comment by tom08zehn** [comment]
I don't know. But my idea was to create a new response transformer and to search for "status" in the "Reference" tab:

![image](https://user-images.githubusercontent.com/12202052/207549288-c645d972-8abc-4cff-bc4d-2f0e9dca0e0f.png)

Have you already tried `responseStatusMessage = 'Patient not found';` ?

Maybe this is "passed through" and used as HTTP status code message?

**Reply by scrumgelist** [reply]
Thanks Tom,

I looked at this I thought, "Is it that easy?  Surely I would have tried this already?"

Maybe I had, but I tried anyway just to be sure - alas no joy.

I put this in my response transformer:

    	       logger.info($('responseStatusLine'));
    	       responseMap.put('responseStatusLine', 'HTTP/1.1 540 Patient not found');
    	       logger.info($('responseStatusLine'));
    	       
First pass the value is blank (the destination is a web service sender not an http sender so no surprise there), then it gets the value I set.  (I wasn't sure about _'responseMap.put'_ but it worked and the value was showing within the 'Response' scope of the dashboard mappings after the message was processed).

---

As an agreed workaround during testing, I'm adding a value-pair into the http response header of "ResponsePhrase":"myShortMessage eg 'Patient not found'", but really need to put it in the first line of the http header for Prod and don't see how to take any significant control of this in Mirth.  I really really am loath to resort to any intercept and repackage shinanigans outside of it.

**Reply by tom08zehn** [reply]
The screenshot above shows that you selected "Pathology Webservice" as the channel's response.

![image](https://user-images.githubusercontent.com/12202052/207716392-6bc77f1b-9666-4d46-887c-9ab26bb8870b.png)

What is it - the name of a destination or a key in your `responseMap` ?

--

The following only applies if "Pathology Webservice" is a destination: Even the Mirth Administrator Guide says that you can override the status message as described above (+ example):

https://www.nextgen.com/-/media/files/nextgen-connect/nextgen-connect-39-user-guide.pdf

![image](https://user-images.githubusercontent.com/12202052/207712588-35734aec-a13c-4dcb-b3b3-2a7de1bf2507.png)

--

If "Pathology Webservice" is a key in your `responseMap` then the `Response` object has a `setStatusMessage()` method:

https://github.com/nextgenhealthcare/connect/blob/master/donkey/src/main/java/com/mirth/connect/donkey/model/message/Response.java#L173

Then maybe you could set it as follows:

```js
responseMap.get('Pathology Webservice').setStatusMessage('Patient not found');
```

**Comment by jonbartels** [comment]
Look at https://github.com/nextgenhealthcare/connect/blob/0974c6ad4d6fc683d544f6f23da0def551812ecc/server/src/com/mirth/connect/connectors/http/HttpReceiver.java#L467

It calls `setStatus(int)`. There is a `setStatus(int, String)` but that is deprecated http://www.servlets.com/javadoc/javax/servlet/http/HttpServletResponse.html#setStatus(int,%20java.lang.String)

So in order for Mirth Connect to set the status line it would have to call the two-argument method.

**Reply by tom08zehn** [reply]
Correct, the problem is [here](https://github.com/nextgenhealthcare/connect/blob/master/server/src/com/mirth/connect/connectors/http/HttpReceiver.java#L451):

```js
            /*
             * If the status code is custom, use the entered/replaced string If is is not a
             * variable, use the status of the destination's response (success = 200, failure = 500)
             * Otherwise, return 200
             */
            if (statusCode != -1) {
                servletResponse.setStatus(statusCode);
            } else if (newMessageStatus != null && newMessageStatus.equals(Status.ERROR)) {
                servletResponse.setStatus(HttpStatus.SC_INTERNAL_SERVER_ERROR);
            } else {
                servletResponse.setStatus(HttpStatus.SC_OK);
            }
```

A HTTP listener _always_ returns as status message

* `200 OK` for status code 200 (success or response status=SENT)
* `500 Server Error` for status code 500 (error or response status=ERROR)
* the string or value of the replaced velocity variable in your connector properties, in your case `540` (resolved from `${customResponseCode}`)

If you would use anything else than numbers then it would fall back again to `200 OK` or `500 Server Error`.

So even if you would be able to modify the Pathology Webservice's HTTP response it wouldn't help because the HTTP Listener would _not_ return its status message and instead create a custom one internally.

You would need to get access to the `HttpServletResponse` and override the status before it is sent. I'm pretty sure you cannot achieve this with pure JavaScript... you'd probably need to rewrite the [`HttpReceiver` class](https://github.com/nextgenhealthcare/connect/blob/master/server/src/com/mirth/connect/connectors/http/HttpReceiver.java#L459)  and modify line 459... or raise a feature request!

---

[Added:] setting the response status message as follows only impacts the Mirth _internal_ response but not the HTTP response:

You can see `responseStatusMessage = 'Patient not found';` in the Mirth dashboard:

![image](https://user-images.githubusercontent.com/12202052/207834465-6f00e48f-75b3-458d-a40b-a3ac2cd434ae.png)

You can achieve the same using `ResponseFactory` (example creates an "Error Response" instead of "Sent Response"):

```js
var r = ResponseFactory.getErrorResponse(''); // empty payload
r.setStatusMessage('Patient not found');
responseMap.put('Pathology Webservice', r);
```

![image](https://user-images.githubusercontent.com/12202052/207835832-dc934d32-22b6-49d0-927f-4a35132455cd.png)

... using the snippets:

![image](https://user-images.githubusercontent.com/12202052/207836294-11a3b26a-ecb6-4005-81da-faefaeb3a57d.png)

Or you could create a custom response from scratch:

```js
// See https://github.com/nextgenhealthcare/connect/blob/master/donkey/src/main/java/com/mirth/connect/donkey/model/message/Response.java#L70
var r = new com.mirth.connect.donkey.model.message.Response(com.mirth.connect.donkey.model.message.Status.SENT, null, 'Patient not found');
responseMap.put('Pathology Webservice', r);
```

**Comment by scrumgelist** [comment, marked as answer]
Tom and Jon,

Thank you so much for investing so much time in this.

What can I say?  At least now I know I wasn't going mad.

I'm not going to implement an enhanced class however enticing the idea, mostly because I don't think I'm up to maintaining it across version changes etc.

As it seems so odd to be able to change the http response status code but then not be able set the accompanying http response status text, I feel a feature request coming.

Thanks again, you are both stars.

**Reply by scrumgelist** [reply]
Change Request: https://github.com/nextgenhealthcare/connect/discussions/5589

**Reply by tom08zehn** [reply]
It was a learning for me so I wanted to share... this way I have it "documented" but for the whole community.

Thanks for the change request. I think you revealed a missing and valuable feature here. Idea to add in order to minimize GUI and behavior changes, the Code could:

* Split the string in "Response Status Code" field by space char
* Do the same as today with the first item of the resulting array (try to convert to int etc.)
* If array length > 1 then use the rest as response status message (joined again with space char)

Example: string "540 Patient not found" results in status code 540 and status message "Patient not found" and both then is returned in the HTTP servlet response

Also, a response transformer should get 2 new reserved variables, similar to `responseStatusMessage`, to be able to do the following (only available/applies in the response transformer of a HTTP Sender connector):

* `httpResponseStatusCode = 540;`
* `httpResponseStatusMessage = "Patient not found"`

**Comment by wi4seo** [comment]
Hi Team,

FREE Registration for Mirth Unconference 2023 is Now Opened!

To know more, please visit the event page and register yourself: https://wi4.org/mirth-connect-unconference-feb-2023/

Stay Tuned for More Updates!​

​Thanks & Regards,
